libvast/src/arrow_table_slice_builder.cpp

```diff
@@ -388,6 +387,63 @@ class map_column_builder : public arrow_table_slice_builder::column_builder {
   std::unique_ptr<column_builder> val_builder_;
 };
 
+class record_column_builder : public arrow_table_slice_builder::column_builder {
+public:
+  // There is no RecordBuilder in Arrow. A record is simply a list of structs
+  // (key-value pairs).
+  using arrow_builder_type = arrow::ListBuilder;
+
+  using data_type = view<record>;
+
+  record_column_builder(
+    arrow::MemoryPool* pool, std::shared_ptr<arrow::DataType> struct_type,
+    std::vector<std::unique_ptr<column_builder>>&& field_builders)
+    : field_builders_{std::move(field_builders)} {
+    auto fields = std::vector<std::shared_ptr<arrow::ArrayBuilder>>{};
+    fields.reserve(field_builders_.size());
+    for (auto& field_builder : field_builders_)
+      fields.push_back(field_builder->arrow_builder());
+    struct_builder_ = std::make_shared<arrow::StructBuilder>(struct_type, pool,
+                                                             std::move(fields));
+    list_builder_ = std::make_shared<arrow::ListBuilder>(pool, struct_builder_);
+  }
+
+  bool add(data_view x) override {
+    if (caf::holds_alternative<view<caf::none_t>>(x))
+      return list_builder_->AppendNull().ok();
+    if (!list_builder_->Append().ok())
+      return false;
+    if (auto* xptr = caf::get_if<data_type>(&x)) {
+      for (auto field : **xptr) {
+        if (!struct_builder_->Append().ok())
+          return false;
+        for (auto& field_builder : field_builders_)
+          if (!field_builder->add(field.second))
+            return false;
+      }
+      return true;
+    }
+    return false;
+  }
+
+  std::shared_ptr<arrow::Array> finish() override {
+    std::shared_ptr<arrow::Array> result;
+    if (!list_builder_->Finish(&result).ok())
+      die("failed to finish Arrow column builder");
+    return result;
+  }
+
+  std::shared_ptr<arrow::ArrayBuilder> arrow_builder() const override {
+    return list_builder_;
+  }
+
+private:
+  std::shared_ptr<arrow::StructBuilder> struct_builder_;
+  std::shared_ptr<arrow_builder_type> list_builder_;
```

Comment:
```suggestion
  std::shared_ptr<arrow::ListBuilder> list_builder_;
```

Only if `Builder::arrow_builder_type` is not a thing.